### PR TITLE
Support setting recommended updates

### PR DESCRIPTION
**Pull request by johnf**
This change is [Reviewable](https://reviewable.io/reviews/dsccommunity/xwindowsupdate/31)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -28,6 +28,7 @@ If no log is used, a temporary log name is created by the resource.
 * **UpdateNow**: Indicates if the resource should trigger an update during consistency (including initial.)
 * **Category**: Indicates the categories (one or more) of updates the resource should update for.  'Security', 'Important', 'Optional'.  Default: 'Security' (please note that security is not mutually exclusive with Important and Optional, so selecting Important may install some security updates, etc.)
 * **Notifications**: Sets the windows update agent notification setting.  Supported options are 'disabled' and 'ScheduledInstallation'.  [Documentation from Windows Update](https://msdn.microsoft.com/en-us/library/windows/desktop/aa385806%28v=vs.85%29.aspx?f=255&MSPPError=-2147217396)
+* **RecommendedUpdates**: Whether to include Recommended updates (default false).
```
Comment (TravisEz13) [inline README.md]: This description is not accurate.  Please update the description.  Please include the impact on `Optional` and `Important` updates.

**Review by TravisEz13 — CHANGES_REQUESTED** [review]

**Comment by TravisEz13** [comment]
If you need help understanding the impact of the change to write the description.  You can find my contact information in my profile.

**Comment by johnf** [comment]
Hi @TravisEz13 could you take another look at this? I've updated the descriptions and variable names in the code

**Review by TravisEz13 — APPROVED** [review]

**Comment by TravisEz13** [comment]
My approval is for your changes.  The tests need to be fixed.

**Comment by johlju** [comment]
@johnf Are you able to work on this issue so we can bring this to the finish line?

**Comment by johlju** [comment]
Labeling this PR as abandoned since it has gone 14 days or more since the last update. An abandoned PR can be continued by another contributor. The abandoned label will be removed if work on the PR is taken up again.

**Comment by johnf** [comment]
OK Tests are now working. Although they are currently broken for the project.

Note the commit in ```appveyor.yml``` is a hack to make the tests pass.

The last commit to PowerShell/DscResource.Tests#230 is breaking the tests. See https://ci.appveyor.com/project/PowerShell/xwindowsupdate/build/2.5.193.0 for an example. Not sure if problem is on our end or theirs.

I've allowed edits from maintainers so you can remove that commit.

**Comment by johlju** [comment]
Sorry about the trouble with the tests! I made review comments to resolved that without reverting to an older commit.

Do you have a chance to resolve these review comments?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[appveyor.yml, line 7 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaC_BU-4PdST5zFhtA:-LDaC_BV3QA5tl4FdoRh:b-cx1put) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/appveyor.yml#L7)):*
> ```YAML
> install: 
>     - git clone https://github.com/PowerShell/DscResource.Tests
>     - cd DscResource.Tests
> ```

For line 7-9 should be able to be removed if resolving the other comments for this file.

---

*[appveyor.yml, line 12 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaCXFJAW4eMBOgD9YZ:-LDaCXFJAW4eMBOgD9Y_:bz8m76) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/appveyor.yml#L12)):*
> ```YAML
>     - ps: |
>         $moduleName = 'xWindowsUpdate'
>         $mainModuleFolder = ".\"
> ```

This should not be needed, see previous comment.

---

*[appveyor.yml, line 30 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaCKsPA4biDu5e_6eS:-LDaCKsPA4biDu5e_6eT:bbxn1tt) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/appveyor.yml#L30)):*
> ```YAML
>         Invoke-AppveyorTestScriptTask `
>             -Type 'Default' `
>             -MainModulePath $mainModuleFolder `
> ```

This parameter should not be here. It is not used when `Type` is `'Default'`.

---

*[README.md, line 31 at r1](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-K_71oeXRerLRzcSeLs8-r1-31:-LDaET1F742z4nPZ5ktm:biyidcp) ([raw file](https://github.com/powershell/xwindowsupdate/blob/9a538cf4654da8443e470a3eb1383ed0d9c42b7c/README.md#L31)):*
<details><summary><i>Previously, TravisEz13 (Travis Plunk) wrote…</i></summary><blockquote>

This description is not accurate.  Please update the description.  Please include the impact on `Optional` and `Important` updates.
</blockquote></details>

This is not part of the code that is changed in this PR, so issue #47 has been submitted for his instead. 

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 179 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaJ_8Q3Yr470CvWd0Z:-LDaJ_8Q3Yr470CvWd0_:b-c0evll) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L179)):*
> ```PowerShell
> }
> 
> function Get-WuaAuIncludeRecommendedUpdates
> ```

Please add comment-based help to this helper function, at least `.SYNOPSIS`.

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 181 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaFwPCD9Toji_paFOU:-LDaFwPCD9Toji_paFOV:b-51et0q) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L181)):*
> ```
> -tryBlock
> ```

`-TryBlock` (upper 'T')

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 250 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaJhDZ4yyfxYeIMTbd:-LDaJhDZ4yyfxYeIMTbe:b-gek8d5) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L250)):*
> ```PowerShell
> }
> 
> function Set-WuaAuIncludeRecommendedUpdates
> ```

Please add comment-based help to this helper function, at least `.SYNOPSIS` and `.PARAMETER`.

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 252 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaG0yj6Y8BsyqawkUb:-LDaG0yj6Y8BsyqawkUc:b-z4dqwg) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L252)):*
> ```PowerShell
> function Set-WuaAuIncludeRecommendedUpdates
> {
>     param(
> ```

parenthesis should be on a separate row.

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 253 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaG6e482dDVEalWVNA:-LDaG6e482dDVEalWVNB:b-gdnlm7) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L253)):*
> ```
> [bool]
> ```

`[Boolean]`

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 367 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaG8qY2TUvbQA8ttCw:-LDaG8qY2TUvbQA8ttCx:b-31kvn1) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L367)):*
> ```
> [bool]
> ```

`[Boolean]`. Throughout this new parameter.

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 474 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaIXwO2wzG6jGuHylN:-LDaIXwO2wzG6jGuHylO:b-34ygjv) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L474)):*
> ```PowerShell
>     Write-Verbose "notifications compliant: $notificationCompliant"
>     $SourceCompliant = (!$Source -or $Source -eq $Get.Source)
>     $recommendedUpdatesCompliant = (!$IncludeRecommendedUpdates -or $IncludeRecommendedUpdates -eq $Get.IncludeRecommendedUpdates)
> ```

See comment in Test-TargetResource for this evaluation.

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 528 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaIxo_8oZ6DZwm26zY:-LDaIxo_8oZ6DZwm26zZ:by212aj) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L528)):*
> ```
> $PSCmdlet.ShouldProcess(
> ```

We should not use ShouldProcess in DSC resources. This can be removed (see there are more of these, but let's not add more to remove to the backlog) :slightly_smiling_face: 

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 532 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaIgVf2efpAIOkQo7A:-LDaIgVf2efpAIOkQo7B:b-kvbxkp) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L532)):*
> ```PowerShell
>                 Try
>                 {
>                     #TODO verify that group policy is not overriding this settings
> ```

Is this todo comment something that needs to resolved prior to merge, or as a separate issue?

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 539 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaJEW0-U8nJRN_IJ9U:-LDaJEW0-U8nJRN_IJ9V:b-wak1df) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L539)):*
> ```PowerShell
>                 {
>                     $ErrorMsg = $_.Exception.Message
>                     Write-Verbose $ErrorMsg
> ```

Shouldn't we throw if there are a problem here? 

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 600 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaHYZS5xldASiJnC4f:-LDaHYZS5xldASiJnC4g:b-mwtoya) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L600)):*
> ```PowerShell
>     Write-Verbose "notifications compliant: $notificationCompliant"
>     $SourceCompliant = (!$Source -or $Source -eq $Get.Source)
>     $recommendedUpdatesCompliant = (!$IncludeRecommendedUpdates -or $IncludeRecommendedUpdates -eq $Get.IncludeRecommendedUpdates)
> ```

Does this return correct if the current state is `$true`, but the user wants `$false`

```powershell
$IncludeRecommendedUpdates = $false
(!$IncludeRecommendedUpdates -or $IncludeRecommendedUpdates -eq $true)
```

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 600 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaHyTq-uxqix0DhO-l:-LDaHyTq-uxqix0DhO-m:b-u4vi00) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L600)):*
> ```PowerShell
>     Write-Verbose "notifications compliant: $notificationCompliant"
>     $SourceCompliant = (!$Source -or $Source -eq $Get.Source)
>     $recommendedUpdatesCompliant = (!$IncludeRecommendedUpdates -or $IncludeRecommendedUpdates -eq $Get.IncludeRecommendedUpdates)
> ```

This parameter should only be tested if it is part of the configuration (`$PSBoundParameters`).

---

*[DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1, line 633 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaIL6fDo9OLg9aDaKe:-LDaIL6fDo9OLg9aDaKf:b-anerpl) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/DscResources/MSFT_xWindowsUpdateAgent/MSFT_xWindowsUpdateAgent.psm1#L633)):*
> ```PowerShell
>         
>         [bool]
>         $IncludeRecommendedUpdates,
> ```

Why do we need to add this parameter if there is no code that uses it? 

---

*[Tests/Unit/MSFT_xWindowsUpdateAgent.tests.ps1, line 118 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaKBbk62YQcE7tuL4G:-LDaKBbk62YQcE7tuL4H:bcgre7u) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/Tests/Unit/MSFT_xWindowsUpdateAgent.tests.ps1#L118)):*
> ```PowerShell
>                 } -Verifiable
> 
>                 Mock Get-WuaAuIncludeRecommendedUpdates -MockWith {
> ```

I'm not seeing this property is tested that it returns the correct value when it is in desired state or not in desired state. We should test the property that it returns the expected result from both Get-TargetResource and Test-TargetResource, and that it call's the expected mocks when it is not in desired state in Set-TargetResource.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-:-LDaCk9EE72vmB67XcWU:b-w0eoc7)*

**Comment by johlju** [comment]
I just merged a fix for the appveyor.yml.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[appveyor.yml, line 7 at r3](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-LDaQY6AErhiTgLNEc4t:-LDaQY6AErhiTgLNEc4u:bu7r7ve) ([raw file](https://github.com/powershell/xwindowsupdate/blob/a54177c451953f5ca3f05103ca2e12671ebe75d1/appveyor.yml#L7)):*
> ```YAML
> install: 
>     - git clone https://github.com/PowerShell/DscResource.Tests
>     - cd DscResource.Tests
> ```

 Line 7-9 should not be necessary now.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xwindowsupdate/31#-:-LDaQbOl2ideBWIXc_3f:b-tj2bpw)*

**Comment by johnf** [comment]
@johlju I don't think I can help much further on this commit. I implemented this code at a previous role where I was doing a small amount of windows automation.
I no longer have access to any systems to do any testing and what I've got so far is stretching my non-existent Powershell knowledge.
Just to make the changes I did this time around I was using appveyor to effectively run the code for me.

Hopefully someone else can pick up the rest of the changes.

**Comment by johlju** [comment]
@johnf Thanks for the work, and thank you for the information! I will label this as abandoned so that someone else in the community can pick this up!

**Pull request closed without merging**